### ISSUE TITLE
Gym duplicate pokemon bug

### ISSUE THREAD

**Issue by walaoaaa1234** [RocketMap/RocketMap#1683]
![unknown](https://cloud.githubusercontent.com/assets/16463054/21748950/76606628-d5cd-11e6-94c3-505123364f8d.png)
~~Happens when all member putting the same pokemon?~~

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Version used: as commit [174967ddab3b54f297eec485da209bc898da9540](https://github.com/RocketMap/RocketMap/commit/174967ddab3b54f297eec485da209bc898da9540)
* Environment name and version (e.g. Python 2.7):2.7.12
* Operating System and version (desktop or mobile):windows 10

**Comment by NHLGames** [comment]
I am getting the same issue.  Truncating all the gym tables seems to fix the problem (for a while) but it seems to come back every few weeks.

**Comment by drallieiv** [comment]
I was going to ask for data, but I just seen on my map : 

![this gym](http://i.imgur.com/daFWbNg.jpg)

**Comment by drallieiv** [comment]
On the database side, shouldn't there be a unicity index ?

You could add it with : 
ALTER TABLE `gymmember` ADD UNIQUE INDEX `gymmember_no_duplicates` (`pokemon_uid` ASC, `gym_id` ASC);

**Comment by drallieiv** [comment]
This is the part that is supposed to clear all current pokemon in a gym before adding the new ones.
```
    with flaskDb.database.transaction():
        # Get rid of all the gym members, we're going to insert new records.
        if gym_details:
            DeleteQuery(GymMember).where(
                GymMember.gym_id << gym_details.keys()).execute()

        # Insert new gym members.
        if gym_members:
            db_update_queue.put((GymMember, gym_members))

    log.info('Upserted gyms: %d, gym members: %d.',
             len(gym_details),
             len(gym_members))
```

**Comment by tokar86a** [comment]
I have notes this type of problem to.

**Comment by kinar** [comment]
I've also been dealing with this issue.  I initially tried truncating the gym tables to add a primary key on the gymmember pokemon_uid column (no pokemon should be in more than one gym at once right?) but that resulted in a lot of gyms getting no detailed info at all.  I've since reverted that change and am back to the duplicates but it seems to get worse every day (not sure how that makes sense because I've been running for almost a week without restart).  I've got ~850 pokes in gyms normally in my databse and it grows at a rate of about 200 per hour.

I had almost 800 pokes in a single gym once.  

For now, the workaround I have is to simply delete the duplicates in the gympokemon table every 10mins or so.

```
DELETE n1 FROM gympokemon n1,
    gympokemon n2 
WHERE
    n1.last_seen < n2.last_seen
    AND n1.pokemon_uid = n2.pokemon_uid
```
This leaves the most recent last_seen entry in the table and deletes all the rest.
